discovery-provider/src/tasks/index_solana_plays.py

```diff
@@ -0,0 +1,305 @@
+                                f"track_id: {track_id} "
+                                f"source: {source} "
+                                f"created_at: {created_at} "
+                                f"slot: {tx_slot} "
+                                f"sig: {tx_sig}")
+
+                    session.add(
+                        Play(
+                            user_id=user_id,
+                            play_item_id=track_id,
+                            created_at=created_at,
+                            source=source,
+                            slot=tx_slot,
+                            signature=tx_sig
+                        )
+                    )
+        else:
+            logger.info(f"index_solana_plays.py | tx={tx_sig} Failed to find SECP_PROGRAM")
+    except Exception as e:
+        logger.error(f"index_solana_plays.py | Error processing {tx_sig}, {e}", exc_info=True)
+
+
+# Query the highest traversed solana slot
+def get_latest_slot(db):
+    latest_slot = None
+    with db.scoped_session() as session:
+        highest_slot_query = (
+            session.query(Play)
+            .filter(Play.slot != None)
+            .filter(Play.signature != None)
+            .order_by(desc(Play.slot))
+        ).first()
+        logger.error(f"index_solana_plays.py | get_latest_slot: {highest_slot_query}")
+        # Can be None prior to first write operations
+        if highest_slot_query is not None:
+            latest_slot = highest_slot_query.slot
+
+    # If no slots have yet been recorded, assume all are valid
+    if latest_slot is None:
+        latest_slot = 0
+
+    logger.info(
+        f"index_solana_plays.py | returning {latest_slot} for highest slot"
+    )
+    return latest_slot
+
+# Query a tx signature and confirm its existence
+def get_tx_in_db(session, tx_sig):
+    logger.info(f"index_solana_plays.py | checking db for {tx_sig}")
+    exists = False
+    tx_sig_db_count = (session.query(Play).filter(
+        Play.signature == tx_sig)).count()
+    exists = tx_sig_db_count > 0
+    logger.info(f"index_solana_plays.py | {tx_sig} exists={exists}")
+    return exists
+
+def process_solana_plays(solana_client):
+    if not TRACK_LISTEN_PROGRAM:
+        logger.info("index_solana_plays.py | No program configured, exiting")
+        return
+
+    db = index_solana_plays.db
+
+    # Highest currently processed slot in the DB
+    latest_processed_slot = get_latest_slot(db)
+    logger.info(
+        f"index_solana_plays.py | latest used slot: {latest_processed_slot}")
+
+    # Loop exit condition
+    intersection_found = False
+
+    # List of signatures that will be populated as we traverse recent operations
+    transaction_signatures = []
+
+    # Current batch of transactions
+    transaction_signature_batch = []
+
+    last_tx_signature = None
+
+    # Traverse recent records until an intersection is found with existing Plays table
+    while not intersection_found:
```

Comment:
is it possible to not find an intersection and spin in here forever?

Reply:
no it is not - an "intersection" is considered found if there are no remaining records as well, that is how each discovery provider can traverse history all the way until the initial program deploy